timemachine/md/align.py

```diff
@@ -0,0 +1,96 @@
+import functools
+from typing import Tuple
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+from numpy.typing import NDArray
+from rdkit import Chem
+
+from timemachine.fe import topology
+from timemachine.fe.utils import get_mol_masses, get_romol_conf
+from timemachine.ff import Forcefield
+from timemachine.integrator import LangevinIntegrator
+from timemachine.potentials import generic
+
+
+def align_mols_by_core(
+    mol_a: Chem.Mol, mol_b: Chem.Mol, core: NDArray, ff: Forcefield, n_steps: int = 100, k: float = 10000
+) -> Tuple[NDArray, NDArray]:
+    """
+    Given two mols and a core mapping, simulate the mols in vacuum with harmonic bounds between them to get aligned poses.
+
+    Returns the poses with the lowest RMSD between the cores.
+
+    Parameters
+    ----------
+    mol_a: Chem.Mol
+        First Ligand.
+
+    mol_b: Chem.Mol
+        Second Ligand.
+
+    core: np.ndarray
+        K x 2 mapping of coordinates in A to B. Assumes core[:, 0] matches mol_a and core[:, 1] matches mol_b
+
+    ff: ff.Forcefield
+        Wrapper class around a list of handlers.
+
+    n_steps: int
+        Number of steps to run vacuum simulation for.
+
+    k: float
+        Harmonic bond K value
+
+    Returns
+    -------
+    tuple (mol_a_coords, mol_b_coords)
+        New coords of the two molecules.
+
+
+    """
+    mol_a_conf = get_romol_conf(mol_a)
+    mol_b_conf = get_romol_conf(mol_b)
+    x0 = np.concatenate([mol_a_conf, mol_b_conf])
+    masses = np.concatenate([get_mol_masses(mol_a), get_mol_masses(mol_b)])
+    bond_indices = np.array(core)
+    bond_indices[:, 1] += mol_a_conf.shape[0]
+    top = topology.DualTopology(mol_a, mol_b, ff)
+
+    vacuum_system = top.setup_end_state()
+    U = vacuum_system.get_U_fn()
+
+    fb = generic.HarmonicBond(bond_indices)
+    restraint_params = np.zeros((len(core), 2))
+    restraint_params[:, 0] = k
+    bond_U = functools.partial(
+        fb.to_reference(),
+        params=jnp.array(restraint_params),
+        box=None,
+        lam=0.0,
+    )
+
+    def combined_U(x):
+        return U(x) + bond_U(x)
+
+    grad_fn = jax.jit(jax.grad(combined_U, argnums=(0)))
+
+    def force_func(x):
+        du_dx = grad_fn(x)
+        return -du_dx
+
+    # TBD 1e-3 step size is arbitrary
+    intg = LangevinIntegrator(force_func, masses, 0.0, 1e-3, 0.0)
+    key = jax.random.PRNGKey(2022)  # seed value doesn't matter with no friction and 0 temp
```

Comment:
Possibility: Could [retain 0 temp but set friction > 0](https://github.com/proteneer/timemachine/blob/05d60b173f92ed18b8c63fbf05cd3249a015711e/tests/test_reference_langevin_integrator.py#L152-L158) to get more minimizer-like behavior.

(With friction=temp=0, this may not effectively minimize energy, I think)

Reply:
Resolved in https://github.com/proteneer/timemachine/pull/813/commits/eba33b2cc34690fd53adc6a7399301203d7f3518

Reply:
On further thought -- is there a strong reason to prefer using LangevinIntegrator settings to emulate a minimizer, instead of using e.g. `scipy.optimize.minimize`?

May be preferable since no settings need to be potentially tuned for stability / performance (no stepsize, masses, or damping parameters). (Would expect to be slightly cheaper too -- probably converging in < 100 gradient evaluations.)

Reply:
Don't think there is a very strong case. I would hesitate to use `scipy.optimize.minimize` due to the issues we saw with trying to minimize using scipy for `minimize_host_4d`, though that is a more complicated case admittedly. Would be comfortable using the fire minimizer though as that didn't exhibit the issues?

Reply:
Hmm, fair enough -- would be in favor of tabling this for now.

I don't remember the details of the case where FIRE appears more robust than L-BFGS, or if any causes are in common with this application... Switching optimizer maybe something to consider benchmarking in a follow-up, since this works as-is and the application doesn't seem to be very sensitive to whether the alignment dynamics is energy-minimizing... 

(Main benefit of the suggestion would be to _reduce_ the amount of investigation required to tune parameters / confirm stability etc. -- would defeat the purpose if this instead required new investigation!)